contracts/external/cw-abc/src/test_tube/integration_tests.rs

```diff
@@ -514,3 +519,151 @@ fn test_update_curve() {
         })
     );
 }
+
+#[test]
+fn test_existing_token_failures() {
+    let app = OsmosisTestApp::new();
+    let builder = TestEnvBuilder::new();
+
+    // The tokenfactory token does not exist - fails in supply query
+    let mut instantiate_msg = InstantiateMsg {
+        fees_recipient: "replaced to accounts[0]".to_string(),
+        supply: SupplyToken {
+            token_info: TokenInfo::Existing {
+                denom: "factory/address/nonexistent".to_string(),
+            },
+            decimals: 6,
+            max_supply: Some(Uint128::from(1000000000u128)),
+        },
+        reserve: ReserveToken {
+            denom: RESERVE.to_string(),
+            decimals: 6,
+        },
+        phase_config: CommonsPhaseConfig {
+            hatch: HatchConfig {
+                contribution_limits: MinMax {
+                    min: Uint128::from(10u128),
+                    max: Uint128::from(1000000u128),
+                },
+                initial_raise: MinMax {
+                    min: Uint128::from(10u128),
+                    max: Uint128::from(1000000u128),
+                },
+                entry_fee: Decimal::percent(10u64),
+                exit_fee: Decimal::percent(10u64),
+            },
+            open: OpenConfig {
+                entry_fee: Decimal::percent(10u64),
+                exit_fee: Decimal::percent(10u64),
+            },
+            closed: ClosedConfig {},
+        },
+        hatcher_allowlist: None,
+        curve_type: CurveType::Constant {
+            value: Uint128::one(),
+            scale: 1,
+        },
+    };
+    let result = builder.setup(&app, instantiate_msg.clone());
+    assert!(result.is_err());
+
+    // A subdenom is required
+    let builder = TestEnvBuilder::new();
+    instantiate_msg.supply.token_info = TokenInfo::Existing {
+        denom: "factory/malformed".to_string(),
+    };
+    let result = builder.setup(&app, instantiate_msg.clone());
+    assert!(result.is_err());
+
+    // Only tokenfactory tokens are supported
+    let builder = TestEnvBuilder::new();
+    instantiate_msg.supply.token_info = TokenInfo::Existing {
+        denom: "junox".to_string(),
+    };
+    let result = builder.setup(&app, instantiate_msg.clone());
+    assert!(result.is_err());
+}
+
+#[test]
+fn test_existing_token() {
```

Comment:
For existing tokens, I think we need more tests to ensure no invariants. For example, if I use an existing token of which I have 1 million tokens, then the curve opens up and there are only enough funds to burn 500k... what happens if I try to sell the full amount? Just get zero for the rest of it?

One of the advantages of bonding curve based tokens is that tokens can't go to zero, but using existing tokens breaks this. Maybe not the end of the world, but we should think of test cases around this if we're going to support it.

Reply:
Also should be queries to know if there was a pre-mint, that's important context for users.

Reply:
My understanding of this was that everyone could still sell, but people would start buying in deeper in the curve during the hatch phase. This would lead to less reserve relative to supply in the Open phase. 

This could be fine from the user-perspective if the donate function is changed to match the logic in the blog post's 1st image: https://medium.com/commonsstack/deep-dive-augmented-bonding-curves-b5ca4fad4436. Donations should be sent to the reserve pool instead of funding. This would increase the reserve relative to supply and - for DAO's with profit being distributed back in - could fix this "ponzi-like" situation.  

Definitely a good idea to be able to show as much context as possible though. I will try to think of some specific test-cases as well.

I also think the current use-cases for the `curve_state.funding` can just be removed and automatically sent to the `fees_recipient`.

Reply:
After studying this some more, it's definitely not a good idea to have an existing token supply. DAO's wanting to use an ABC for liquidity should start as a membership-based DAO to instantiate the cw-abc and then update their voting module. I'm going to revert some of these changes here.